### PR TITLE
feat: Added support for `tfupdate` to update version constraints in Terraform configurations

### PR DESCRIPTION
**Pull request by jrottenberg**
Put an `x` into the box if that apply:

- [ ] This PR introduces breaking change.
- [ ] This PR fixes a bug.
- [x] This PR adds new functionality.
- [ ] This PR enhances existing functionality.

### Description of your changes

Fixes https://github.com/antonbabenko/pre-commit-terraform/issues/328

### How can we test changes

Working on an existing module, for example : https://github.com/cloudposse/terraform-aws-sso.git

Prepare the pre-commit config : 

```yaml
repos: # pre-commit autoupdate
  - repo: https://github.com/jrottenberg/pre-commit-terraform
    rev: tfupdate
    hooks:
      - id: tfupdate
        name: tfupdate terraform
      - id: tfupdate
        name: tfupdate provider aws
        args:
          - provider
          - aws   
```

```
❯ pre-commit run -a
[...]
tfupdate terraform.......................................................Failed
- hook id: tfupdate
- files were modified by this hook
tfupdate provider aws....................................................Failed
- hook id: tfupdate
- files were modified by this hook
```

And for the diff :

```
diff --git a/examples/complete/versions.tf b/examples/complete/versions.tf
index 42f8195..764880f 100644
--- a/examples/complete/versions.tf
+++ b/examples/complete/versions.tf
@@ -1,5 +1,5 @@
 terraform {
-  required_version = ">= 0.13.0"
+  required_version = "1.1.5"
 
   required_providers {
     local = "~> 1.2"
diff --git a/modules/account-assignments/versions.tf b/modules/account-assignments/versions.tf
index c9a3bb8..f470932 100644
--- a/modules/account-assignments/versions.tf
+++ b/modules/account-assignments/versions.tf
@@ -1,10 +1,10 @@
 terraform {
-  required_version = ">= 0.13.0"
+  required_version = "1.1.5"
 
   required_providers {
     aws = {
       source  = "hashicorp/aws"
-      version = ">= 3.26.0"
+      version = "4.0.0"
     }
   }
```

teraform version and the provider are pinned.

On subsequent run :

```
❯ pre-commit run -a
[...]
tfupdate terraform.......................................................Passed
tfupdate provider aws....................................................Passed
```

### CODE REVIEW

**Comment by jrottenberg** [comment]
Beside the new feature I had to path  the `__init__.py` file since the warning for terraform_docs_replace was leaking on any python hook.

tfupdate uses subcommands no options it was easier to write the hook in python.

File: .pre-commit-hooks.yaml [hunk 1]
```diff
@@ -111,3 +111,12 @@
...
   exclude: \.terraform\/.*$
   require_serial: true
+
+- id: tfupdate
+  name: tfupdate
+  description: Runs tfupdate on Terraform templates.
+  language: python
+  entry: hook_tfupdate
+  require_serial: true
+  always_run: true
```
Comment (yermulnik) [inline .pre-commit-hooks.yaml:121]: Since `tfupdate` is sort of intrusive, I'd expect the new feature to be opt-in rather than opt-out.

File: README.md [hunk 2]
```diff
@@ -25,27 +25,29 @@ If you are using `pre-commit-terraform` already or want to support its developme
...
+    - [terraform_docs_replace (deprecated)](#terraform_docs_replace-deprecated)
+    - [terraform_fmt](#terraform_fmt)
+    - [terraform_providers_lock](#terraform_providers_lock)
+    - [terraform_tflint](#terraform_tflint)
+    - [terraform_tfsec](#terraform_tfsec)
+    - [terraform_validate](#terraform_validate)
+    - [terrascan](#terrascan)
+    - [tfupdate](#tfupdate)
+  - [Authors](#authors)
+  - [License](#license)
```
Comment (yermulnik) [inline README.md:50]: What is the reason behind this change?

File: hooks/tfupdate.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import subprocess
```
Comment (yermulnik) [inline hooks/tfupdate.py:1]: So what you offer is to invoke Python to run a tool intended to be invoked from within shell and hence add Python as run dependency for a tiny shell tool. Bear in mind that consumer who invokes pre-commit-terraform does this from within shell already. And the only reasoning you provide is the easiness of writing this in Python for you, whereas most of the already existing hooks in this repo are basically only a dozen lines of shell code. I'm a bit confused to be honest.
I see this as opt-in shell-based hook.

Comment (yermulnik) [reply hooks/tfupdate.py:1]: Also in general it's worth to check the code with at least one code linter:
```
tmp/test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
tmp/test.py:5:0: C0116: Missing function or method docstring (missing-function-docstring)
tmp/test.py:10:4: C0103: Variable name "procArgs" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
tmp/test.py:14:4: C0103: Variable name "procArgs" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
tmp/test.py:21:4: C0103: Variable name "e" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
tmp/test.py:5:9: W0613: Unused argument 'argv' (unused-argument)


Report
======
21 statements analysed.


-----------------------------------
Your code has been rated at 7.14/10
```

File: README.md (hunk 2 above)
Comment (jrottenberg) [reply README.md:50]: Vscode going above and beyond, I'll revert

File: hooks/tfupdate.py (hunk 3 above)
Comment (jrottenberg) [reply hooks/tfupdate.py:1]: That's fair, I can write it in bash, just the passing of args from pre-commit was not intuitive at first but I can look closer for sure.

Comment (yermulnik) [reply hooks/tfupdate.py:1]: Thank you.

File: .pre-commit-hooks.yaml (hunk 1 above)
Comment (jrottenberg) [reply .pre-commit-hooks.yaml]: Updated with https://github.com/antonbabenko/pre-commit-terraform/pull/342/commits/2c6c813cbabe75b1fa4585748f1ae28f6c0d4cfa

File: README.md (hunk 2 above)
Comment (jrottenberg) [reply README.md]: reverted

File: hooks/tfupdate.py (hunk 3 above)
Comment (jrottenberg) [reply hooks/tfupdate.py]: All bash, left the python code alone

**Comment by jrottenberg** [comment]
@yermulnik @MaxymVlasov , ok updated the PR as recommended, let me know if it needs anything else.

**Review by MaxymVlasov — COMMENTED** [review]
Let's use common functions to parse args and then use `function per_dir_hook_unique_part` and `function run_hook_on_whole_repo` as replace for `function tfupdate_`


First should work only for dirs with changes, second - recursive for whole repo

Check https://github.com/antonbabenko/pre-commit-terraform/blob/master/hooks/terrascan.sh as an example

**Comment by jrottenberg** [comment]
> Let's use common functions to parse args and then use `function per_dir_hook_unique_part` and `function run_hook_on_whole_repo` as replace for `function tfupdate_`
> 
> First should work only for dirs with changes, second - recursive for whole repo
> 
> Check https://github.com/antonbabenko/pre-commit-terraform/blob/master/hooks/terrascan.sh as an example

tfupdate behaves differently than terrascan, it really runs only against the module as a whole, not individual files. It makes sure the version of terraform or the various providers is pinned to the latest when tfupdate ran. 

The version lookup is a bit expansive and I'd go against running it multiple times, as it will give the same value, but could be considered s abuse from github (it uses the api to extract the latest published version).


I'll make the change to use `common::parse_cmdline`

**Comment by jrottenberg** [comment]
Ok it is consistent with other hooks in term of passing options but seems a bit verbose args of args, when the command itself (tfupdate) takes only a subcommands and parameters. Either way, let me know if you prefer this version, I'll squash and push.

**Comment by jrottenberg** [comment]
ok @MaxymVlasov I'm going with the flow, let me know if I'm missing anything else.

Thanks for the feedback.

**Review by MaxymVlasov — COMMENTED** [review]
LGTM. Just need to check that all works fine. I'll do that little bit later

**Comment by jrottenberg** [comment]
Argh, I'll fix the pre-commit.

File: Dockerfile [hunk 4]
```diff
@@ -126,6 +128,16 @@ RUN . /.env && \
     ) && chmod +x tfsec \
     ; fi
 
+# TFUpdate
+RUN . /.env && \
+    if [ "$TFUPDATE_VERSION" != "false" ]; then \
+    ( \
+        TFUPDATE_RELEASES="https://api.github.com/repos/minamijoyo/tfupdate/releases" && \
+        [ "$TFUPDATE_VERSION" = "latest" ] && curl -L "$(curl -s ${TFUPDATE_RELEASES}/latest | grep -o -E "https://.+?/tfupdate_.+_linux_amd64.tar.gz")" > tfupdate.tgz \
+        || curl -L "$(curl -s ${TFUPDATE_RELEASES} | grep -o -E "https://.+?/v${TF_VERSION}/tfupdate_${TF_VERSION}_linux_amd64.tar.gz")" > tfupdate.tgz \
```
Comment (yermulnik) [inline Dockerfile:137]:
```suggestion
        || curl -L "$(curl -s ${TFUPDATE_RELEASES} | grep -o -E "https://[^,]+?/v${TFUPDATE_VERSION}/tfupdate_${TFUPDATE_VERSION}_linux_amd64.tar.gz")" > tfupdate.tgz \
```

File: Dockerfile [hunk 5]
```diff
@@ -138,6 +150,7 @@ RUN . /.env && \
     (if [ "$TERRASCAN_VERSION"      != "false" ]; then echo "terrascan $(./terrascan version)" >> $F; else echo "terrascan SKIPPED" >> $F      ; fi) && \
     (if [ "$TFLINT_VERSION"         != "false" ]; then ./tflint --version >> $F;                      else echo "tflint SKIPPED" >> $F         ; fi) && \
     (if [ "$TFSEC_VERSION"          != "false" ]; then echo "tfsec $(./tfsec --version)" >> $F;       else echo "tfsec SKIPPED" >> $F          ; fi) && \
+    (if [ "$TFUPDATE_VERSION"       != "false" ]; then echo "tfsec $(./tfupdate --version)" >> $F;    else echo "tfupdate SKIPPED" >> $F       ; fi) && \
```
Comment (yermulnik) [inline Dockerfile:153]:
```suggestion
    (if [ "$TFUPDATE_VERSION"       != "false" ]; then echo "tfupdate $(./tfupdate --version)" >> $F;    else echo "tfupdate SKIPPED" >> $F       ; fi) && \
```

File: README.md [hunk 6]
```diff
@@ -132,6 +136,7 @@ curl -L "$(curl -s https://api.github.com/repos/aquasecurity/tfsec/releases/late
 curl -L "$(curl -s https://api.github.com/repos/accurics/terrascan/releases/latest | grep -o -E -m 1 "https://.+?_Linux_x86_64.tar.gz")" > terrascan.tar.gz && tar -xzf terrascan.tar.gz terrascan && rm terrascan.tar.gz && sudo mv terrascan /usr/bin/ && terrascan init
 sudo apt install -y jq && \
 curl -L "$(curl -s https://api.github.com/repos/infracost/infracost/releases/latest | grep -o -E -m 1 "https://.+?-linux-amd64.tar.gz")" > infracost.tgz && tar -xzf infracost.tgz && rm infracost.tgz && sudo mv infracost-linux-amd64 /usr/bin/infracost && infracost register
+curl -L "$(curl -s https://api.github.com/repos/minamijoyo/tfupdate/releases/latest | grep -o -E -m 1 "https://.+?/tfupdate_.+_linux_amd64.tar.gz")" > tfupdate.tar.gz && tar -xzf tfupdate.tar.gz tfupdate && rm tfupdate.tar.gz && sudo mv tfupdate /usr/bin/
```
Comment (yermulnik) [inline README.md:139]:
```suggestion
curl -L "$(curl -s https://api.github.com/repos/minamijoyo/tfupdate/releases/latest | grep -o -E -m 1 "https://.+?_.+_linux_amd64.tar.gz")" > tfupdate.tar.gz && tar -xzf tfupdate.tar.gz tfupdate && rm tfupdate.tar.gz && sudo mv tfupdate /usr/bin/
```

File: README.md [hunk 7]
```diff
@@ -151,6 +156,7 @@ curl -L "$(curl -s https://api.github.com/repos/terraform-linters/tflint/release
 curl -L "$(curl -s https://api.github.com/repos/aquasecurity/tfsec/releases/latest | grep -o -E -m 1 "https://.+?tfsec-linux-amd64")" > tfsec && chmod +x tfsec && sudo mv tfsec /usr/bin/
 sudo apt install -y jq && \
 curl -L "$(curl -s https://api.github.com/repos/infracost/infracost/releases/latest | grep -o -E -m 1 "https://.+?-linux-amd64.tar.gz")" > infracost.tgz && tar -xzf infracost.tgz && rm infracost.tgz && sudo mv infracost-linux-amd64 /usr/bin/infracost && infracost register
+curl -L "$(curl -s https://api.github.com/repos/minamijoyo/tfupdate/releases/latest | grep -o -E -m 1 "https://.+?/tfupdate_.+_linux_amd64.tar.gz")" > tfupdate.tar.gz && tar -xzf tfupdate.tar.gz tfupdate && rm tfupdate.tar.gz && sudo mv tfupdate /usr/bin/
```
Comment (yermulnik) [inline README.md:159]:
```suggestion
curl -L "$(curl -s https://api.github.com/repos/minamijoyo/tfupdate/releases/latest | grep -o -E -m 1 "https://.+?_.+_linux_amd64.tar.gz")" > tfupdate.tar.gz && tar -xzf tfupdate.tar.gz tfupdate && rm tfupdate.tar.gz && sudo mv tfupdate /usr/bin/
```

File: hooks/tfupdate.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+#   If failed - print out hook checks status
+#######################################################################
+function per_dir_hook_unique_part {
+  local -r args="$1"
+  # shellcheck disable=SC2034 # Unused var.
+  local -r dir_path="$2"
+
+  # pass the arguments to hook
+  # shellcheck disable=SC2068 # hook fails when quoting is used ("$arg[@]")
+  tfupdate "${args[@]}" "${dir_path}"
```
Comment (yermulnik) [inline hooks/tfupdate.sh:35]: Looking at other hooks, shouldn't ti be like the below? Otherwise `tfupdate` will parse `${args[@]}` as single command line arg, imho.
```suggestion
  tfupdate ${args[@]} "${dir_path}"
```

File: hooks/tfupdate.sh [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+# Unique part of `common::per_dir_hook`. The function is executed one time
+# in the root git repo
+# Arguments:
+#   args (string with array) arguments that configure wrapped tool behavior
+#######################################################################
+function run_hook_on_whole_repo {
+  local -r args="$1"
+  # pass the arguments to hook
+  # shellcheck disable=SC2068 # hook fails when quoting is used ("$arg[@]")
+  tfupdate "${args[*]}" --recursive .
```
Comment (yermulnik) [inline hooks/tfupdate.sh:52]: Looking at other hooks, shouldn't ti be like the below? Otherwise `tfupdate` will parse `${args[@]}` as single command line arg, imho.
```suggestion
  tfupdate ${args[*]} --recursive .
```

**Comment by jrottenberg** [comment]
Thank you very much @yermulnik , I was missing the targe on the "tar x".

Let me know what you think @MaxymVlasov

**Comment by jrottenberg** [comment]
@MaxymVlasov very strange, the pre-commit are passing locally, hadolint included.

**Comment by jrottenberg** [comment]
rebased and ready for review

**Review by MaxymVlasov — APPROVED** [review]
Great job!
I fixed all things where I had some questions, so now LGTM

Let's wait to merge #362, rebase here and if all pipelines will be happy and @yermulnik will approve PR, we can go to merge it

File: README.md [hunk 10]
```diff
@@ -620,6 +626,33 @@ Example:
...
+        - --args=provider aws # Will be pined to latest version
+
+    - id: tfupdate
+      name: Autoupdate Helm provider versions
+      args:
+        - --args=provider helm
+        - --args=--version 2.5.0 # Will be pined to specified version
+    ```
+
+Check [`tfupdate` usage instructio](https://github.com/minamijoyo/tfupdate#usage) for other available options and usage examples.
```
Comment (yermulnik) [inline README.md:653]:
```suggestion
Check [`tfupdate` usage instructions](https://github.com/minamijoyo/tfupdate#usage) for other available options and usage examples.  
```

**Review by yermulnik — APPROVED** [review]
LGTM

**Comment by antonbabenko** [comment]
This PR is included in version 1.66.0 :tada:

**Comment by jrottenberg** [comment]
Thank you !
